vowpalwabbit/global_data.h

```diff
@@ -34,96 +37,77 @@
 #include "example.h"
 #include "config.h"
 #include "learner.h"
-#include "v_hashmap.h"
 #include <time.h>
 #include "hash.h"
 #include "crossplat_compat.h"
 #include "error_reporting.h"
 #include "constant.h"
 #include "rand48.h"
+#include "hashstring.h"
 
 #include "options.h"
 #include "version.h"
 
 typedef float weight;
 
-typedef v_hashmap<substring, features*> feature_dict;
+typedef std::unordered_map<std::string, std::unique_ptr<features>> feature_dict;
 
 struct dictionary_info
 {
-  char* name;
+  std::string name;
   uint64_t file_hash;
-  feature_dict* dict;
+  std::shared_ptr<feature_dict> dict;
 };
 
-inline void deleter(substring ss, uint64_t /* label */) { free_it(ss.begin); }
-
 class namedlabels
 {
  private:
-  std::vector<substring> id2name;
-  v_hashmap<substring, uint64_t> name2id;
-  uint32_t K;
+  // NOTE: This ordering is critical. m_id2name and m_name2id contain pointers into m_label_list!
+  std::string m_label_list;
```

Comment:
[Minor][Nit] You add "m_" prefixes in this file, but in others you add "_" prefixes.

Reply:
We had a discussion late into the development of this branch to prefix with _ instead of m_. I'll leave this as-is for now, but we should make a sweep through the code to fix up the variable names at some point